mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb

```diff
@@ -1303,11 +1302,35 @@ func resourceSqlDatabaseInstanceUpdate(d *schema.ResourceData, meta interface{})
 		return err
 	}
 
-	// Update only updates the settings, so they are all we need to set.
-	instance := &sqladmin.DatabaseInstance{
-		Settings: expandSqlDatabaseInstanceSettings(d.Get("settings").([]interface{})),
+	desiredSetting := d.Get("settings")
+	var op *sqladmin.Operation
+	var instance *sqladmin.DatabaseInstance
+	// Major version upgrade
+	if v, ok := d.GetOk("database_version"); ok {
+		instance = &sqladmin.DatabaseInstance{DatabaseVersion: v.(string)}
+		err = retryTimeDuration(func() (rerr error) {
+			op, rerr = config.NewSqlAdminClient(userAgent).Instances.Patch(project, d.Get("name").(string), instance).Do()
+			return rerr
+		}, d.Timeout(schema.TimeoutUpdate), isSqlOperationInProgressError)
+		if err != nil {
+			return fmt.Errorf("Error, failed to patch instance settings for %s: %s", instance.Name, err)
+		}
+		err = sqlAdminOperationWaitTime(config, op, project, "Patch Instance", userAgent, d.Timeout(schema.TimeoutUpdate))
+		if err != nil {
+			return err
+		}
+		err = resourceSqlDatabaseInstanceRead(d, meta)
+		if err != nil {
+			return err
+		}
 	}
 
+	s := d.Get("settings")
+	instance = &sqladmin.DatabaseInstance{
+		Settings: expandSqlDatabaseInstanceSettings(desiredSetting.([]interface{})),
+	}
+	_settings := s.([]interface{})[0].(map[string]interface{})
+	instance.Settings.SettingsVersion = int64(_settings["version"].(int))
```

Comment:
I assume this is due to a change in the API package? This kind of translation should be handled inside expandSqlDatabaseInstanceSettings. You might also need to do the reverse translation inside flattenSqlDatabaseInstanceSettings.

Reply:
Hi Stephen, this is because after the patch completes for `database_version`, the setting proto version has been updated by +8. This fails the Instance.Update request as the old version in the state file doesn't match the new version post the Instance.Patch request. 

The resourceSqlDatabaseInstanceCreate operation is doing something very similar here https://github.com/GoogleCloudPlatform/magic-modules/pull/6374/files/b7a43e3c5614771620a5ce4fdc93522b8f639658#diff-b34dfa25708e844d450cdb21f989baa23ffe466f2b817267606b10adc69048c1R893

Reply:
Ah, I see what you mean - since you (might) call `resourceSqlDatabaseInstanceRead` above, that overrides the Terraform state for the resource with the current settings version; this is manually extracting just that information and updating the instance object that will be used for the Update request. (created from a pre-Patch version of the settings - so correct desired end state but outdated version).

Could you add a comment explaining what you're doing?

Reply:
Added comment explaining the situation.